fe/fe-core/src/main/java/com/starrocks/alter/OptimizeJobV2.java

```diff
@@ -0,0 +1,653 @@
+        this.tmpPartitionIds = tmpPartitionIds;
+    }
+
+    public String getName() {
+        return "optimize-" + this.postfix;
+    }
+
+    public Map<String, String> getProperties() {
+        return properties;
+    }
+
+    public List<OptimizeTask> getOptimizeTasks() {
+        return rewriteTasks;
+    }
+
+    private Database getAndReadLockDatabase(long dbId) throws AlterCancelException {
+        Database db = GlobalStateMgr.getCurrentState().getDb(dbId);
+        if (db == null) {
+            throw new AlterCancelException("database id: " + dbId + " does not exist");
+        }
+        if (!db.readLockAndCheckExist()) {
+            throw new AlterCancelException("insert overwrite commit failed because locking db: " + dbId + " failed");
+        }
+        return db;
+    }
+
+    private OlapTable checkAndGetTable(Database db, long tableId) throws AlterCancelException {
+        Table table = db.getTable(tableId);
+        if (table == null) {
+            throw new AlterCancelException("table: " + tableId + " does not exist in database: " + db.getFullName());
+        }
+        Preconditions.checkState(table instanceof OlapTable);
+        return (OlapTable) table;
+    }
+
+    private Database getAndWriteLockDatabase(long dbId) throws AlterCancelException {
+        Database db = GlobalStateMgr.getCurrentState().getDb(dbId);
+        if (db == null) {
+            throw new AlterCancelException("database id:" + dbId + " does not exist");
+        }
+        if (!db.writeLockAndCheckExist()) {
+            throw new AlterCancelException("insert overwrite commit failed because locking db:" + dbId + " failed");
+        }
+        return db;
+    }
+
+    /**
+     * runPendingJob():
+     * 1. Create all temp partitions and wait them finished.
+     * 2. Get a new transaction id, then set job's state to WAITING_TXN
+     */
+    @Override
+    protected void runPendingJob() throws AlterCancelException {
+        Preconditions.checkState(jobState == JobState.PENDING, jobState);
+
+        LOG.info("begin to send create temp partitions. job: {}", jobId);
+        Database db = GlobalStateMgr.getCurrentState().getDb(dbId);
+        if (db == null) {
+            throw new AlterCancelException("Database " + dbId + " does not exist");
+        }
+
+        if (!checkTableStable(db)) {
+            return;
+        }
+
+        // 1. create temp partitions
+        for (int i = 0; i < optimizeClause.getSourcePartitionIds().size(); ++i) {
+            tmpPartitionIds.add(GlobalStateMgr.getCurrentState().getNextId());
+        }
+
+        long createPartitionStartTimestamp = System.currentTimeMillis();
+        OlapTable targetTable;
+        db.readLock();
+        try {
+            targetTable = checkAndGetTable(db, tableId);
+        } finally {
+            db.readUnlock();
+        }
+        try {
+            PartitionUtils.createAndAddTempPartitionsForTable(db, targetTable, postfix,
```

Comment:
1. ```create table t(k int) distributed by hash(k)``` will change the bucket num or not?
2. During the optimization, does the new ingestion fail or reject?
3.  Could the colocate table bucket be optimized?

Reply:
1. it will
2. optimize will fail when partition has new ingestion in default
3. after optimize will be not colocate